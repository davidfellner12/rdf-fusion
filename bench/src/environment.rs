use crate::BenchmarkingOptions;
use crate::benchmarks::BenchmarkName;
use crate::prepare::{PrepRequirement, prepare_run_closure, prepare_run_command};
use crate::prepare::{ensure_file_download, prepare_file_download};
use anyhow::bail;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

/// Represents a context used to execute benchmarks.
pub struct RdfFusionBenchContext {
    /// General options for the benchmarks.
    options: BenchmarkingOptions,
    /// The path to the data dir.
    data_dir: Mutex<PathBuf>,
    /// The path to the current directory. This might be different from the `data_dir` if a
    /// benchmark is running.
    results_dir: Mutex<PathBuf>,
}

impl RdfFusionBenchContext {
    /// Creates a new [RdfFusionBenchContext].
    pub fn new(
        options: BenchmarkingOptions,
        data_dir: PathBuf,
        results_dir: PathBuf,
    ) -> Self {
        Self {
            options,
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(results_dir),
        }
    }

    /// Returns the [BenchmarkingOptions] for this context.
    pub fn options(&self) -> &BenchmarkingOptions {
        &self.options
    }

    /// Resolves a relative path `file` against the data directory.
    pub fn join_data_dir(&self, file: &Path) -> anyhow::Result<PathBuf> {
        if !file.is_relative() {
            bail!("Only relative paths can be resolved.")
        }

        Ok(self.data_dir.lock().unwrap().join(file))
    }

    /// Creates a new folder in the results directory and uses it until [Self::pop_dir] is
    /// called.
    ///
    /// This can be used to create folder hierarchies to separate the results of different
    /// benchmarks.
    #[allow(clippy::create_dir)]
    pub fn push_dir(&self, dir: &str) -> anyhow::Result<()> {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.push(dir);
        results_dir.push(dir);

        Ok(())
    }

    /// Pops the last directory from the stack.
    pub fn pop_dir(&self) {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.pop();
        results_dir.pop();
    }

    /// Creates a new bencher and modifies the context for this benchmark.
    pub fn create_benchmark_context(
        &self,
        benchmark_name: BenchmarkName,
    ) -> anyhow::Result<BenchmarkContext<'_>> {
        self.push_dir(&benchmark_name.dir_name())?;
        Ok(BenchmarkContext {
            context: self,
            benchmark_name,
        })
    }
}

/// A benchmarker that can be used to execute benchmarks.
///
/// It holds a reference to the current context to store its results.
pub struct BenchmarkContext<'ctx> {
    /// Reference to the benchmarking context.
    context: &'ctx RdfFusionBenchContext,
    /// Name of the benchmark that is being executed.
    benchmark_name: BenchmarkName,
}

impl<'ctx> BenchmarkContext<'ctx> {
    /// Returns a reference to the benchmarking context.
    pub fn parent(&self) -> &RdfFusionBenchContext {
        self.context
    }

    /// Returns the name of the benchmark that is being executed.
    pub fn benchmark_name(&self) -> BenchmarkName {
        self.benchmark_name
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn data_dir(&self) -> PathBuf {
        self.context.data_dir.lock().unwrap().clone()
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn results_dir(&self) -> PathBuf {
        self.context.results_dir.lock().unwrap().clone()
    }

    /// Prepares the context such that `requirement` is fulfilled.
    pub async fn prepare_requirement(
        &self,
        requirement: PrepRequirement,
    ) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload {
                url,
                file_name,
                action,
            } => prepare_file_download(self, url, file_name, action).await,
            PrepRequirement::RunClosure { execute, .. } => {
                prepare_run_closure(self, &execute)
            }
            PrepRequirement::RunCommand {
                workdir,
                program,
                args,
                ..
            } => {
                let workdir = self.context.join_data_dir(&workdir)?;
                prepare_run_command(&workdir, &program, &args)
            }
        }
    }

    /// Ensures that the `requirement` is fulfilled in this context.
    pub fn ensure_requirement(&self, requirement: PrepRequirement) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload { file_name, .. } => {
                ensure_file_download(self, file_name.as_path())
            }
            PrepRequirement::RunClosure {
                check_requirement, ..
            }
            | PrepRequirement::RunCommand {
                check_requirement, ..
            } => check_requirement(self),
        }
    }
}

/// Pops the results directory from the context when the bencher is dropped.
impl Drop for BenchmarkContext<'_> {
    fn drop(&mut self) {
        self.context.pop_dir();
    }
}
