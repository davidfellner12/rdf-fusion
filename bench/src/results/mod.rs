mod benchmark_results;
mod run;

pub use benchmark_results::*;
pub use run::*;
