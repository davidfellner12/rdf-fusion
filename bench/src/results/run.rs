use anyhow::Context;
use serde::{Deserialize, Serialize};
use std::fmt::{Display, Formatter};
use std::time::Duration;

/// Contains the runs of a single benchmark.
#[derive(Debug, Serialize, Deserialize)]
pub struct BenchmarkRuns {
    /// The individual runs.
    runs: Vec<BenchmarkRun>,
}

impl BenchmarkRuns {
    /// Creates a new empty [BenchmarkRuns].
    pub fn new() -> Self {
        Self { runs: Vec::new() }
    }

    /// Adds a new [BenchmarkRun].
    pub fn add_run(&mut self, run: BenchmarkRun) {
        self.runs.push(run);
    }

    pub fn summarize(&self) -> anyhow::Result<BenchmarkSummary> {
        let number_of_samples = u32::try_from(self.runs.len())
            .context("Too many samples for computing the average duration")?;
        let avg_duration =
            self.runs.iter().map(|run| run.duration).sum::<Duration>() / number_of_samples;

        Ok(BenchmarkSummary {
            number_of_samples,
            avg_duration,
        })
    }
}

impl Default for BenchmarkRuns {
    fn default() -> Self {
        Self::new()
    }
}

/// Represents a single run of a benchmark.
#[derive(Debug, Serialize, Deserialize)]
pub struct BenchmarkRun {
    /// The duration of the benchmark run.
    pub duration: Duration,
}

/// Holds the statistics computed over [BenchmarkRuns].
#[derive(Debug, Serialize, Deserialize)]
pub struct BenchmarkSummary {
    /// Represents how often the benchmark was executed.
    pub number_of_samples: u32,
    /// The average duration over all benchmark runs.
    pub avg_duration: Duration,
}

impl Display for BenchmarkSummary {
    #[allow(clippy::use_debug)]
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Benchmark Summary:")?;
        writeln!(f, "Number of Samples: {}", self.number_of_samples)?;
        write!(f, "Average Duration: {:?}", self.avg_duration)
    }
}
