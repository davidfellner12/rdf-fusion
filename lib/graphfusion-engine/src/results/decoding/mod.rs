use crate::results::decoding::logical::DecodeRdfTermsNode;
use crate::DFResult;
use arrow_rdf::encoded::EncTerm;
use datafusion::arrow::datatypes::{Field, Schema, SchemaRef};
use datafusion::logical_expr::{Extension, LogicalPlan};
use std::sync::Arc;

mod logical;
mod rewrite;

pub use rewrite::DecodeRdfTermsToProjectionRule;

pub fn decode_rdf_terms(input: LogicalPlan) -> DFResult<LogicalPlan> {
    Ok(LogicalPlan::Extension(Extension {
        node: Arc::new(DecodeRdfTermsNode::new(input)?),
    }))
}

fn compute_decoded_schema(input_schema: &Schema) -> SchemaRef {
    let fields: Vec<_> = input_schema
        .fields()
        .iter()
        .map(|f| transform_field(f))
        .collect();
    SchemaRef::new(Schema::new(fields))
}

fn transform_field(field: &Field) -> Field {
    if *field.data_type() == EncTerm::term_type() {
        Field::new(field.name(), EncTerm::term_type(), field.is_nullable())
    } else {
        field.clone()
    }
}
