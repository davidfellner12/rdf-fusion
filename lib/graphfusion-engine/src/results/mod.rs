use arrow_rdf::decoded::model::DecTerm;
use arrow_rdf::decoded::DecRdfTermBuilder;
use datafusion::arrow::array::RecordBatch;
use datafusion::arrow::datatypes::{Field, Schema, SchemaRef};
use datafusion::arrow::error::ArrowError;
use datafusion::error::DataFusionError;
use datafusion::physical_plan::memory::MemoryStream;
use futures::StreamExt;
use oxrdf::{Variable, VariableRef};
use oxrdfio::{RdfFormat, RdfSerializer};
use sparesults::{
    QueryResultsFormat, QueryResultsParseError, QueryResultsParser, QueryResultsSerializer,
    ReaderQueryResultsParserOutput,
};
use std::error::Error;
use std::io::{Read, Write};
use std::sync::Arc;

mod decoding;
mod graph_name;
mod quads;
mod query_solution;
mod triples;

use crate::sparql::error::EvaluationError;
pub use decoding::decode_rdf_terms;
pub use decoding::DecodeRdfTermsToProjectionRule;
pub use graph_name::GraphNameStream;
pub use quads::QuadStream;
pub use query_solution::QuerySolutionStream;
pub use sparesults::QuerySolution;
pub use triples::QueryTripleStream;

/// Results of a [SPARQL query](https://www.w3.org/TR/sparql11-query/).
pub enum QueryResults {
    /// Results of a [SELECT](https://www.w3.org/TR/sparql11-query/#select) query.
    Solutions(QuerySolutionStream),
    /// Result of a [ASK](https://www.w3.org/TR/sparql11-query/#ask) query.
    Boolean(bool),
    /// Results of a [CONSTRUCT](https://www.w3.org/TR/sparql11-query/#construct) or [DESCRIBE](https://www.w3.org/TR/sparql11-query/#describe) query.
    Graph(QueryTripleStream),
}

impl QueryResults {
    /// Reads a SPARQL query results serialization.
    pub fn read(
        reader: impl Read + 'static,
        format: QueryResultsFormat,
    ) -> Result<Self, QuerySolutionsToStreamError> {
        let parser = QueryResultsParser::from_format(format).for_reader(reader)?;
        query_result_for_parser(parser)
    }

    /// Writes the query results (solutions or boolean).
    ///
    /// This method fails if it is called on the `Graph` results.
    ///
    /// ```
    /// use graphfusion::store::Store;
    /// use graphfusion::model::*;
    /// use graphfusion::sparql::results::QueryResultsFormat;
    ///
    /// let store = Store::new()?;
    /// let ex = NamedNodeRef::new("http://example.com")?;
    /// store.insert(QuadRef::new(ex, ex, ex, GraphNameRef::DefaultGraph))?;
    ///
    /// let results = store.query("SELECT ?s WHERE { ?s ?p ?o }")?;
    /// assert_eq!(
    ///     results.write(Vec::new(), QueryResultsFormat::Json)?,
    ///     r#"{"head":{"vars":["s"]},"results":{"bindings":[{"s":{"type":"uri","value":"http://example.com"}}]}}"#.as_bytes()
    /// );
    /// # Result::<_, Box<dyn std::error::Error>>::Ok(())
    /// ```
    pub async fn write<W: Write>(
        self,
        writer: W,
        format: QueryResultsFormat,
    ) -> Result<W, EvaluationError> {
        let serializer = QueryResultsSerializer::from_format(format);
        match self {
            Self::Boolean(value) => serializer.serialize_boolean_to_writer(writer, value),
            Self::Solutions(mut solutions) => {
                let mut serializer = serializer
                    .serialize_solutions_to_writer(writer, solutions.variables().to_vec())
                    .map_err(EvaluationError::ResultsSerialization)?;
                while let Some(solution) = solutions.next().await {
                    serializer
                        .serialize(&solution?)
                        .map_err(EvaluationError::ResultsSerialization)?;
                }
                serializer.finish()
            }
            Self::Graph(triples) => {
                let s = VariableRef::new_unchecked("subject");
                let p = VariableRef::new_unchecked("predicate");
                let o = VariableRef::new_unchecked("object");
                let mut serializer = serializer
                    .serialize_solutions_to_writer(
                        writer,
                        vec![s.into_owned(), p.into_owned(), o.into_owned()],
                    )
                    .map_err(EvaluationError::ResultsSerialization)?;
                for triple in triples {
                    let triple = triple?;
                    serializer
                        .serialize([
                            (s, &triple.subject.into()),
                            (p, &triple.predicate.into()),
                            (o, &triple.object),
                        ])
                        .map_err(EvaluationError::ResultsSerialization)?;
                }
                serializer.finish()
            }
        }
        .map_err(EvaluationError::ResultsSerialization)
    }

    /// Writes the graph query results.
    ///
    /// This method fails if it is called on the `Solution` or `Boolean` results.
    ///
    /// ```
    /// use graphfusion::io::RdfFormat;
    /// use graphfusion::model::*;
    /// use graphfusion::store::Store;
    ///
    /// let graph = "<http://example.com> <http://example.com> <http://example.com> .\n";
    ///
    /// let store = Store::new()?;
    /// store.load_graph(
    ///     graph.as_bytes(),
    ///     RdfFormat::NTriples,
    ///     GraphName::DefaultGraph,
    ///     None,
    /// )?;
    ///
    /// let results = store.query("CONSTRUCT WHERE { ?s ?p ?o }")?;
    /// assert_eq!(
    ///     results.write_graph(Vec::new(), RdfFormat::NTriples)?,
    ///     graph.as_bytes()
    /// );
    /// # Result::<_, Box<dyn std::error::Error>>::Ok(())
    /// ```
    pub fn write_graph<W: Write>(
        self,
        writer: W,
        format: impl Into<RdfFormat>,
    ) -> Result<W, EvaluationError> {
        if let Self::Graph(triples) = self {
            let mut serializer = RdfSerializer::from_format(format.into()).for_writer(writer);
            for triple in triples {
                serializer
                    .serialize_triple(&triple?)
                    .map_err(EvaluationError::ResultsSerialization)?;
            }
            serializer
                .finish()
                .map_err(EvaluationError::ResultsSerialization)
        } else {
            Err(EvaluationError::NotAGraph)
        }
    }
}

/// Indicates that there was a problem while turning a query result into a query solution stream.
#[derive(Debug, thiserror::Error)]
pub enum QuerySolutionsToStreamError {
    #[error("There was an error while obtaining the query solutions")]
    QuerySolutionSource(#[from] Box<dyn Error + Send + Sync>),
    #[error("Could not create a record batch from the result")]
    RecordBatchCreation(#[from] ArrowError),
    #[error("Could not create a stream from the resulting record batch")]
    StreamCreation(#[from] DataFusionError),
}

impl From<QueryResultsParseError> for QuerySolutionsToStreamError {
    fn from(value: QueryResultsParseError) -> Self {
        Self::QuerySolutionSource(Box::new(value))
    }
}

fn query_result_for_parser(
    parser: ReaderQueryResultsParserOutput<impl Read + Sized>,
) -> Result<QueryResults, QuerySolutionsToStreamError> {
    Ok(match parser {
        ReaderQueryResultsParserOutput::Solutions(s) => {
            let variables: Arc<[Variable]> = s.variables().into();
            let parser_iter = s
                .into_iter()
                .map(|r| r.map_err(|error| Box::new(error) as Box<dyn Error + Send + Sync>));
            query_result_for_iterator(variables, parser_iter)?
        }
        ReaderQueryResultsParserOutput::Boolean(v) => QueryResults::Boolean(v),
    })
}

pub fn query_result_for_iterator(
    variables: Arc<[Variable]>,
    solutions: impl Iterator<Item = Result<QuerySolution, Box<dyn Error + Send + Sync>>>,
) -> Result<QueryResults, QuerySolutionsToStreamError> {
    let mut builders = Vec::new();
    for _ in 0..variables.len() {
        builders.push(DecRdfTermBuilder::new())
    }

    for solution in solutions {
        let solution =
            solution.map_err(|err| QuerySolutionsToStreamError::QuerySolutionSource(err))?;
        for (idx, (_, term)) in solution.iter().enumerate() {
            builders
                .get_mut(idx)
                .expect("Initialized with enough builders")
                .append_term(term)?
        }
    }

    let fields = variables
        .iter()
        .map(|v| Field::new(v.as_str(), DecTerm::term_type(), true))
        .collect::<Vec<_>>();
    let columns = builders
        .into_iter()
        .map(|builder| builder.finish())
        .collect::<Result<Vec<_>, ArrowError>>()?;

    let schema = SchemaRef::new(Schema::new(fields));
    let record_batch = RecordBatch::try_new(schema.clone(), columns)?;
    let record_batch_stream = MemoryStream::try_new(vec![record_batch], schema, None)?;
    let stream = QuerySolutionStream::new(variables, Box::pin(record_batch_stream));
    Ok(QueryResults::Solutions(stream))
}

impl From<QuerySolutionStream> for QueryResults {
    #[inline]
    fn from(value: QuerySolutionStream) -> Self {
        Self::Solutions(value)
    }
}
