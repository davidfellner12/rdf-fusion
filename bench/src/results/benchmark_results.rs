use crate::benchmarks::BenchmarkName;
use crate::results::run::BenchmarkRuns;
use crate::results::BenchmarkRun;
use std::collections::HashMap;

/// Represents the overall results of running (multiple) benchmarks.
pub struct BenchmarkResults {
    /// A mapping from the executed benchmark to all recorded runs.
    runs: HashMap<BenchmarkName, BenchmarkRuns>,
}

impl BenchmarkResults {
    /// Returns a new empty [BenchmarkResults].
    pub fn new() -> Self {
        Self {
            runs: HashMap::new(),
        }
    }

    /// Returns the runs of a given benchmark.
    pub fn get_runs(&self, name: BenchmarkName) -> Option<&BenchmarkRuns> {
        self.runs.get(&name)
    }

    /// Adds a new run for a given `namer
    pub fn add_run(&mut self, name: BenchmarkName, run: BenchmarkRun) {
        let runs = self.runs.entry(name).or_insert(BenchmarkRuns::new());
        runs.add_run(run);
    }
}
