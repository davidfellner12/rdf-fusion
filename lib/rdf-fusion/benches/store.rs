#![allow(clippy::panic)]

use bzip2::read::MultiBzDecoder;
use codspeed_criterion_compat::{criterion_group, criterion_main, Criterion, Throughput};
use futures::StreamExt;
use oxrdfio::RdfFormat;
use rdf_fusion::store::Store;
use rdf_fusion_engine::results::QueryResults;
use rdf_fusion_engine::sparql::{Query, QueryOptions};
use reqwest::Url;
use std::fs::File;
use std::io::{Read, Write};
use std::path::Path;
use std::str;
use std::str::FromStr;
use tokio::runtime::Runtime;

fn store_load(c: &mut Criterion) {
    let data = read_bz2_data("https://zenodo.org/records/12663333/files/dataset-1000.nt.bz2");
    let mut group = c.benchmark_group("store load");
    group.throughput(Throughput::Bytes(data.len() as u64));
    group.sample_size(10);
    group.bench_function("load BSBM explore 1000 in memory", |b| {
        b.to_async(Runtime::new().unwrap()).iter(|| async {
            let store = Store::new();
            do_load(&store, &data).await;
        });
    });
}

async fn do_load(store: &Store, data: &[u8]) {
    store
        .load_from_reader(RdfFormat::NTriples, data)
        .await
        .unwrap();
}

fn store_query_and_update(c: &mut Criterion) {
    for data_size in [1_000, 5_000] {
        do_store_query_and_update(c, data_size)
    }
}

fn do_store_query_and_update(c: &mut Criterion, data_size: usize) {
    let data = read_bz2_data(&format!(
        "https://zenodo.org/records/12663333/files/dataset-{data_size}.nt.bz2"
    ));
    let explore_operations = bsbm_sparql_operation("exploreAndUpdate-1000.csv.bz2")
        .into_iter()
        .filter_map(|op| match op {
            RawOperation::Query(q) => {
                // TODO remove once describe is supported
                if q.contains("DESCRIBE") {
                    None
                } else {
                    Some(Operation::Query(Query::parse(&q, None).unwrap()))
                }
            }
            RawOperation::Update(_) => None,
        })
        .collect::<Vec<_>>();
    let explore_query_operations = explore_operations
        .iter()
        .filter(|o| matches!(o, Operation::Query(_)))
        .cloned()
        .collect::<Vec<_>>();
    let business_operations = bsbm_sparql_operation("businessIntelligence-1000.csv.bz2")
        .into_iter()
        .filter_map(|op| match op {
            RawOperation::Query(q) => {
                // TODO remove once describe is supported
                if q.contains("DESCRIBE") {
                    None
                } else {
                    Some(Operation::Query(
                        Query::parse(&q.replace('#', ""), None).unwrap(),
                    ))
                }
            }
            RawOperation::Update(_) => unreachable!(),
        })
        .collect::<Vec<_>>();

    let mut group = c.benchmark_group("store operations");
    group.sample_size(10);

    {
        let runtime = Runtime::new().unwrap();
        let memory_store = Store::new();
        runtime.block_on(do_load(&memory_store, &data));

        group.bench_function(format!("BSBM explore {data_size} query in memory"), |b| {
            b.to_async(Runtime::new().unwrap())
                .iter(|| run_operation(&memory_store, &explore_query_operations));
        });
        group.bench_function(
            format!("BSBM explore {data_size} queryAndUpdate in memory"),
            |b| {
                b.to_async(Runtime::new().unwrap())
                    .iter(|| run_operation(&memory_store, &explore_operations));
            },
        );
        group.bench_function(
            format!("BSBM business intelligence {data_size} in memory"),
            |b| {
                b.to_async(Runtime::new().unwrap())
                    .iter(|| run_operation(&memory_store, &business_operations));
            },
        );
    }
}

async fn run_operation(store: &Store, operations: &[Operation]) {
    let options = QueryOptions::default();
    for operation in operations {
        match operation {
            Operation::Query(q) => match store.query_opt(q.clone(), options.clone()).await.unwrap()
            {
                QueryResults::Boolean(_) => (),
                QueryResults::Solutions(mut s) => {
                    while let Some(s) = s.next().await {
                        s.unwrap();
                    }
                }
                QueryResults::Graph(mut g) => {
                    while let Some(t) = g.next().await {
                        t.unwrap();
                    }
                }
            },
            // Operation::Update(u) => store.update_opt(u.clone(), options.clone()).await.unwrap(),
        }
    }
}

criterion_group!(store, store_query_and_update, store_load);

criterion_main!(store);

fn read_bz2_data(url: &str) -> Vec<u8> {
    let url = Url::from_str(url).unwrap();
    let file_name = url.path().split('/').next_back().unwrap().to_owned();

    if !Path::new(&file_name).exists() {
        let client = reqwest::blocking::Client::new();
        let response = client.get(url.clone()).send().unwrap();
        assert!(response.status().is_success(), "{}", &url);

        File::create(&file_name)
            .unwrap()
            .write_all(&response.bytes().unwrap())
            .unwrap();
    }

    let mut buf = Vec::new();
    MultiBzDecoder::new(File::open(&file_name).unwrap())
        .read_to_end(&mut buf)
        .unwrap();
    buf
}

fn bsbm_sparql_operation(file_name: &str) -> Vec<RawOperation> {
    csv::Reader::from_reader(read_bz2_data(&format!("https://zenodo.org/records/12663333/files/{file_name}")).as_slice()).records()
        .collect::<Result<Vec<_>, _>>().unwrap()
        .into_iter()
        .rev()
        .take(300) // We take only 10 groups
        .map(|l| {
            match &l[1] {
                "query" => RawOperation::Query(l[2].into()),
                "update" => RawOperation::Update(l[2].into()),
                _ => panic!("Unexpected operation kind {}", &l[1]),
            }
        })
        .collect()
}

#[allow(dead_code)]
#[derive(Clone)]
enum RawOperation {
    Query(String),
    Update(String),
}

#[allow(clippy::large_enum_variant, clippy::allow_attributes)]
#[derive(Clone)]
enum Operation {
    Query(Query),
}
