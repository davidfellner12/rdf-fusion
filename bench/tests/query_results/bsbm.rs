//! This file contains tests for an adapte version of the BSBM explore use case. The adaption have
//! been made such that the queries produce stable results.
//!
//! The results have been compared against GraphDB 11.0.1. The following curl command can be used to
//! get the results of GraphDB. Ideally, one can pipe the result of the request into a file or
//! directly to the clipboard using `| xclip -selection clipboard` (or similar). You can also
//! download the JSON result from the GraphDB UI.
//!
//! ```bash
//! curl -X POST \
//!     "http://<graphdb_url>/repositories/bsbm" \
//!     -H "Content-Type: application/sparql-query" \
//!     -H "Accept: application/sparql-results+json" \
//!     --data-binary '<query>'
//! ```
//!
//! Then, even though we pretty-print our results, there will be some differences (e.g., spacing,
//! order of keys). You can use a tool that semantically compares JSON files to "quickly" check
//! the results of a new test (e.g., [JSON Compare](https://jsoncompare.org/)). `CONSTRUCT` queries
//! have been compared manually.

use futures::StreamExt;
use insta::assert_snapshot;
use rdf_fusion::io::{RdfFormat, RdfSerializer};
use rdf_fusion::results::{QueryResultsFormat, QueryResultsSerializer};
use rdf_fusion::store::Store;
use rdf_fusion::QueryResults;
use rdf_fusion_bench::benchmarks::bsbm::NumProducts::N1_000;
use rdf_fusion_bench::benchmarks::bsbm::{BsbmBenchmark, ExploreUseCase};
use rdf_fusion_bench::benchmarks::Benchmark;
use rdf_fusion_bench::environment::RdfFusionBenchContext;
use serde_json::Value;
use std::path::PathBuf;

#[tokio::test]
pub async fn bsbm_1000_test_results() {
    let benchmarking_context =
        RdfFusionBenchContext::new_for_criterion(PathBuf::from("./data"));
    let benchmark = BsbmBenchmark::<ExploreUseCase>::try_new(N1_000, None).unwrap();
    let benchmark_name = benchmark.name();
    let ctx = benchmarking_context
        .create_benchmark_context(benchmark_name)
        .unwrap();

    let store = benchmark.prepare_store(&ctx).await.unwrap();

    assert_snapshot!(
        "Explore Q1",
        run_select_query(&store, include_str!("./queries/explore-q1.sparql")).await
    );
    assert_snapshot!(
        "Explore Q2",
        run_select_query(&store, include_str!("./queries/explore-q2.sparql")).await
    );
    assert_snapshot!(
        "Explore Q3",
        run_select_query(&store, include_str!("./queries/explore-q3.sparql")).await
    );
    assert_snapshot!(
        "Explore Q5",
        run_select_query(&store, include_str!("./queries/explore-q5.sparql")).await
    );
    assert_snapshot!(
        "Explore Q7",
        run_select_query(&store, include_str!("./queries/explore-q7.sparql")).await
    );
    assert_snapshot!(
        "Explore Q8",
        run_select_query(&store, include_str!("./queries/explore-q8.sparql")).await
    );
    assert_snapshot!(
        "Explore Q10",
        run_select_query(&store, include_str!("./queries/explore-q10.sparql")).await
    );
    assert_snapshot!(
        "Explore Q11",
        run_select_query(&store, include_str!("./queries/explore-q11.sparql")).await
    );
    assert_snapshot!(
        "Explore Q12",
        run_construct_query(&store, include_str!("./queries/explore-q12.sparql")).await
    );
}

async fn run_select_query(store: &Store, query: &str) -> String {
    let result = store.query(query).await.unwrap();
    let QueryResults::Solutions(mut solutions) = result else {
        panic!("Unexpected result format!")
    };

    let mut buffer = Vec::new();
    let mut serializer = QueryResultsSerializer::from_format(QueryResultsFormat::Json)
        .serialize_solutions_to_writer(&mut buffer, solutions.variables().to_vec())
        .unwrap();
    while let Some(solution) = solutions.next().await {
        let solution = solution.unwrap();
        serializer.serialize(solution.iter()).unwrap();
    }
    serializer.finish().unwrap();

    let raw_json = String::from_utf8(buffer).unwrap();
    let v: Value = serde_json::from_str(&raw_json).unwrap();
    serde_json::to_string_pretty(&v).unwrap()
}

async fn run_construct_query(store: &Store, query: &str) -> String {
    let result = store.query(query).await.unwrap();
    let QueryResults::Graph(mut solutions) = result else {
        panic!("Unexpected result format!")
    };

    let mut buffer = Vec::new();
    let mut serializer =
        RdfSerializer::from_format(RdfFormat::Turtle).for_writer(&mut buffer);
    while let Some(solution) = solutions.next().await {
        let solution = solution.unwrap();
        serializer.serialize_triple(solution.as_ref()).unwrap();
    }
    serializer.finish().unwrap();

    String::from_utf8(buffer).unwrap()
}
