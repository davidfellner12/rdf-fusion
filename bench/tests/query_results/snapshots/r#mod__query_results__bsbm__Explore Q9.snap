---
source: bench/tests/query_results/bsbm.rs
expression: "run_graph_result_query(&store,\ninclude_str!(\"./queries/explore-q9.sparql\")).await"
---
<http://www4.wiwiss.fu-berlin.de/bizer/bsbm/v01/instances/dataFromRatingSite1/Reviewer207> a <http://xmlns.com/foaf/0.1/Person> .
